src/apps/applib/include/utility.h

```diff
@@ -42,4 +42,8 @@ void geoBoundaryPrint(const GeoBoundary* b);
 void geoBoundaryPrintln(const GeoBoundary* b);
 int readBoundary(FILE* f, GeoBoundary* b);
 
+void iterateAllIndexesAtRes(int res, void (*callback)(H3Index));
+void iterateAllIndexesAtResPartial(int res, void (*callback)(H3Index),
```

Comment:
Again, C users would probably prefer macros here and you'll get better performance because C compiler can see what is going on in a macro, but function pointers are opaque. Side note: this is why using the C `qsort` in C++ is slower than `std::sort` from C++ STL, due to function pointers being worse off than templates.

Anyway, here's how I'd do it based on [Jansson](https://jansson.readthedocs.io/en/2.11/apiref.html#c.json_array_foreach):

```
#define H3_FOR_EACH_INDEX_RES(argName, res) H3_FOR_EACH_INDEX_PARTIAL((res), NUM_BASE_CELLS)
#define H3_FOR_EACH_INDEX_PARTIAL(argName, res, baseCells) \
    assert((baseCells) <= NUM_BASE_CELLS); \
     for (int i = 0; i < baseCells; i++) { \
         H3Index bc; \
         setH3Index(&bc, 0, i, 0); \
         int childrenSz = H3_EXPORT(maxUncompactSize)(&bc, 1, (res)); \
         STACK_ARRAY_CALLOC(H3Index, children, childrenSz); \
         H3_EXPORT(uncompact)(&bc, 1, children, childrenSz, (res)); \
         for (int j = 0; j < childrenSz; j++) { \
             if (children[j] == 0) { \
                 continue; \
             } \
             H3Index argName = children[j];
```

Then users can do this:

```
H3_FOR_EACH_INDEX_AT_RES(1, myIndex) {
    printf("%lld\n", myIndex);
}
```

Which expands to:

```
for (...) {
    ...
    H3Index myIndex = children[j];
    printf("%lld\n", myIndex);
}
```

Reply:
I'll change this, but I am not a huge fan of macros. Since this is a test, I don't think performance is that big a concern.

Reply:
Ya everyone hates macros until they program in C for a while and realize it's the only way to avoid seriously ugly code. Also, underrated technique: [X macro](https://en.wikipedia.org/wiki/X_Macro).

Meanwhile, rethinking this API too so that the blocks have matching parentheses.

Reply:
OK seeing as this is a test I don't really think it's worth the effort. Thought this might lead to a convenience API in general. Never mind my suggestion.